pkg/upload/upload.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+const (
+	multiPartFormData = "multipart/form-data"
+	EncryptHeader     = "swarm-encrypt"
+)
+
+// FileUploadInfo contains the data for a file to be uploaded
+type FileUploadInfo struct {
+	FileName    string
+	FileSize    int64
+	ContentType string
+	ToEncrypt   bool
+	Reader      io.Reader
+}
+
+// GetFileHTTPInfo extracts file info for upload from HTTP request
+func GetFileHTTPInfo(r *http.Request) (*FileUploadInfo, error) {
+	toEncrypt := strings.ToLower(r.Header.Get(EncryptHeader)) == "true"
+	contentType := r.Header.Get("Content-Type")
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return nil, fmt.Errorf("parse content type error: %v", err)
+	}
+
+	var reader io.Reader
+	var fileName, contentLength string
+	var fileSize uint64
+
+	if mediaType == multiPartFormData {
+		mr := multipart.NewReader(r.Body, params["boundary"])
+
+		// read only the first part, as only one file upload is supported
+		part, err := mr.NextPart()
+		if err != nil {
+			return nil, fmt.Errorf("read multipart error: %v", err)
+		}
+
+		// try to find filename
+		// 1) in part header params
+		// 2) as formname
+		// 3) file reference hash (after uploading the file)
+		if fileName = part.FileName(); fileName == "" {
+			fileName = part.FormName()
+		}
+
+		// then find out content type
+		contentType = part.Header.Get("Content-Type")
+		if contentType == "" {
+			br := bufio.NewReader(part)
+			buf, err := br.Peek(512)
+			if err != nil && err != io.EOF {
+				return nil, fmt.Errorf("read content type error: %v", err)
+			}
+			contentType = http.DetectContentType(buf)
+			reader = br
+		} else {
+			reader = part
+		}
+		contentLength = part.Header.Get("Content-Length")
+	} else {
+		fileName = r.URL.Query().Get("name")
+		contentLength = r.Header.Get("Content-Length")
+		reader = r.Body
+	}
+
+	if contentLength != "" {
+		fileSize, err = strconv.ParseUint(contentLength, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid content length error: %v", err)
+		}
+	} else {
+		// copy the part to a tmp file to get its size
+		tmp, err := ioutil.TempFile("", "bee-multipart")
+		if err != nil {
+			return nil, fmt.Errorf("create temp file error: %v", err)
+		}
+		defer os.Remove(tmp.Name())
+		defer tmp.Close()
+		n, err := io.Copy(tmp, reader)
+		if err != nil {
+			return nil, fmt.Errorf("write temp file error: %v", err)
+		}
+		if _, err := tmp.Seek(0, io.SeekStart); err != nil {
+			return nil, fmt.Errorf("seek temp file error: %v", err)
+		}
+		fileSize = uint64(n)
+		reader = tmp
+	}
+
+	return &FileUploadInfo{
+		FileName:    fileName,
+		FileSize:    int64(fileSize),
+		ContentType: contentType,
+		ToEncrypt:   toEncrypt,
+		Reader:      reader,
+	}, nil
+}
+
+// StoreFile stores the given file and returns its reference
+func StoreFile(ctx context.Context, fileInfo *FileUploadInfo, s storage.Storer) (swarm.Address, error) {
```

Comment:
extracted from the current implementation of `pkg/api/file.go/fileUploadHandler`